### ISSUE TITLE
PodInfo should be a list, not a map

### ISSUE THREAD

**Issue by abonas** [GoogleCloudPlatform/kubernetes#3622]
Comparing a single container details in "spec" and the "status" sections, 
there are some differences in what data is exposed per container in those sections:

1. container is identified with index array in "spec", and its name is a property inside the index.
while in "status" its main element is its name.

2. in the "status" which is the runtime online representation, there's no info about ports, memory, cpu, etc.
even when container is running.

![podports](https://cloud.githubusercontent.com/assets/6277245/5817371/e271ebb2-a0b3-11e4-9489-1be70b43cb8d.png)

**Comment by erictune** [comment]
@bgrant0607 PTAL

**Comment by abonas** [comment]
@erictune what is PTAL?

**Comment by erictune** [comment]
PTAL = "please take a look" or "please take another look".

**Comment by thockin** [comment]
As per docs/api-conventions.md "the convention is to use a list of
subobjects containing name fields"

https://github.com/GoogleCloudPlatform/kubernetes/issues/2004

It sounds like status is buggy

**Comment by abonas** [comment]
but what about the other issue - missing info on ports/resources?

**Comment by bgrant0607** [comment]
Yes, status should not be a map.

**Comment by bgrant0607** [comment]
@abonas What info would you like to see about ports?

We plan to introduce a separate API for resource information. See the section about usage here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/resources.md

/cc @rjnagal

**Comment by abonas** [comment]
@bgrant0607 when I query for pod information and its containers, it's important to see what's going on with them in runtime. so if spec was ports XYZ, cpu A and memory B, I would like to see if indeed it matches the spec or is different. were the ports allocated? same ports? different ports?  
otherwise, having only the spec information, there's nothing user can do to monitor/manage the runtime.

**Comment by abonas** [comment]
also, having an additional api means it's requiring more requests to server, more merging work and resolving relationships on client side. not very convenient for clients.

**Comment by abonas** [comment]
I will open a separate issue on the ports missing info, and keep this one for the buggy status structure

**Comment by piosz** [comment]
@dchen1107 are you working on it?

**Comment by dchen1107** [comment]
not at this moment, feel free grabbing it.

**Comment by piosz** [comment]
@bgrant0607 
I'm trying to change PodInfo to be a list.

Should it only be changed in api v3 or also in v1 and v2? In general do we still want to modify api v1 and v2 or are those versions final?

How about internal representation of PodInfo? Do we want to keep it as it is or change to be compatible with external type? There is some internal code (pkg/kubelet/dockertools/docker.go) which relies on that PodInfo is a map. Do we want to try use externally and internally the same types when possible?

**Comment by thockin** [comment]
For consistency in beta3, I still endorse this.  Brian is contemplating the
longer term of maps vs lists, and this may yet reverse.  I don't see that
happening in v1.

I would make this change only in b3 and internal.  If keeping it as a map
is important internally, we CAN do that, but it requires a conversion func,
which I know Brian is trying to avoid.

**Comment by bgrant0607** [comment]
I recommend waiting on this until at least Monday.

**Comment by piosz** [comment]
Thanks for the info.

@bgrant0607 ping me when you're ready.

**Comment by bgrant0607** [comment]
Sorry for the delay. Hold off on this a bit longer. Debate is ongoing. I recommend choosing something else to work on in the short term.

**Comment by piosz** [comment]
No problem. Sure, I'm working also on other issues. Please let me know the decision will be made.

**Comment by bgrant0607** [comment]
Ok, decision is lists: #4889.

Sorry for the delay.

**Comment by piosz** [comment]
Thanks for the info.

**Comment by thockin** [comment]
I did not get the sense that a decision had been made and I've been
watching that thread...

**Comment by brendandburns** [comment]
I don't think that this makes the v1.0 cut.

**Comment by bgrant0607** [comment]
We were trying to get this into v1beta3, since it currently violates our stated API convention.

That said, I'm fine with demoting the priority at this point, esp. since it's in status rather than spec.

**Comment by piosz** [comment]
I'm about to finish this change (still need to fix one test)
